retrieval/discovery/azure.go

```diff
@@ -0,0 +1,185 @@
+type AzureDiscovery struct {
+	cfg      *config.AzureSDConfig
+	interval time.Duration
+	port     int
+}
+
+// NewAzureDiscovery returns a new AzureDiscovery which periodically refreshes its targets.
+func NewAzureDiscovery(cfg *config.AzureSDConfig) *AzureDiscovery {
+	// creds := credentials.NewStaticCredentials(conf.AccessKey, conf.SecretKey, "")
+	// if conf.AccessKey == "" && conf.SecretKey == "" {
+	// 	creds = defaults.DefaultChainCredentials
+	// }
+	return &AzureDiscovery{
+		cfg:      cfg,
+		interval: time.Duration(cfg.RefreshInterval),
+		port:     cfg.Port,
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (ad *AzureDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(ad.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	tg, err := ad.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := ad.refresh()
+			if err != nil {
+				log.Error(err)
+			} else {
+				ch <- []*config.TargetGroup{tg}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// azureClient represents multiple Azure Resource Manager providers.
+type azureClient struct {
+	nic network.InterfacesClient
+	vm  compute.VirtualMachinesClient
+}
+
+// createAzureClient is a helper function for creating an Azure compute client to ARM.
+func createAzureClient(cfg config.AzureSDConfig) (azureClient, error) {
+	var c azureClient
+	oauthConfig, err := azure.PublicCloud.OAuthConfigForTenant(cfg.TenantID)
+	if err != nil {
+		return azureClient{}, err
+	}
+	spt, err := azure.NewServicePrincipalToken(*oauthConfig, cfg.ClientID, cfg.ClientSecret, azure.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return azureClient{}, err
+	}
+
+	c.vm = compute.NewVirtualMachinesClient(cfg.SubscriptionID)
+	c.vm.Authorizer = spt
+
+	c.nic = network.NewInterfacesClient(cfg.SubscriptionID)
+	c.nic.Authorizer = spt
+
+	return c, nil
+}
+
+func (ad *AzureDiscovery) refresh() (*config.TargetGroup, error) {
+	tg := &config.TargetGroup{}
+	client, err := createAzureClient(*ad.cfg)
+	if err != nil {
+		return tg, fmt.Errorf("could not create Azure client: %s", err)
+	}
+
+	var machines []compute.VirtualMachine
+	result, err := client.vm.ListAll()
+	if err != nil {
+		return tg, fmt.Errorf("could not list virtual machines: %s", err)
+	}
+	machines = append(machines, *result.Value...)
+	// if we still have more results, keep going
+	for result.NextLink != nil {
+		result, err = client.vm.ListAllNextResults(result)
+		if err != nil {
+			return tg, fmt.Errorf("could not list virtual machines: %s", err)
+		}
+		machines = append(machines, *result.Value...)
+	}
+	log.Infof("Found %d virtual machines during Azure discovery.", len(machines))
+
+	// We have the slice of machines. Now turn them into targets.
+	// Doing them in go routines because the network interface calls are slow.
```

Comment:
That's quite a low limit, a 5m default sounds sane - but that's going to still eat a lot of API calls for the user.

This is very limiting on the part of Azue, we should ask them to either bump that substantially or have the list all machines return all the leaf information.